src/graph/impl/KokkosGraph_ExplicitCoarsening_impl.hpp

```diff
@@ -0,0 +1,303 @@
+          lno_t rowDeg = rowmap(row + 1) - rowmap(row);
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(t, rowDeg),
+            [&] (const lno_t j)
+            {
+              lno_t nei = colinds(rowmap(row) + j);
+              //Remote neighbors are not included
+              if(nei >= numRows)
+                return;
+              lno_t neiCluster = vertClusters(nei);
+              if(neiCluster != cluster)
+              {
+                //Have a neighbor. Try to find it in the table.
+                if(!lookup(neiCluster, table))
+                {
+                  //Not in the table. Try to insert it.
+                  insert(cluster, neiCluster, table);
+                  //Whether or not insertion succeeded,
+                  //this is a cross-cluster edge possibly not seen before
+                  mask.set(rowmap(row) + j);
+                }
+              }
+            });
+        });
+    }
+
+    size_t team_shmem_size(int teamSize) const
+    {
+      return tableSize() * sizeof(int);
+    }
+
+    lno_t numRows;
+    fine_rowmap_t rowmap;
+    fine_entries_t colinds;
+    ordinal_view_t clusterOffsets;
+    ordinal_view_t clusterVerts;
+    labels_t vertClusters;
+    bitset_t mask;
+  };
+
+  struct FillClusterEntriesFunctor
+  {
+    FillClusterEntriesFunctor(
+        const fine_rowmap_t& rowmap_, const fine_entries_t& colinds_, const coarse_rowmap_t& clusterRowmap_, const coarse_entries_t& clusterEntries_, const ordinal_view_t& clusterOffsets_, const ordinal_view_t& clusterVerts_, const labels_t& vertClusters_, const bitset_t& edgeMask_)
+      : rowmap(rowmap_), colinds(colinds_), clusterRowmap(clusterRowmap_), clusterEntries(clusterEntries_), clusterOffsets(clusterOffsets_), clusterVerts(clusterVerts_), vertClusters(vertClusters_), edgeMask(edgeMask_)
+    {}
+    //Run this scan over entries in clusterVerts (reordered point rows)
+    KOKKOS_INLINE_FUNCTION void operator()(const lno_t i, lno_t& lcount, const bool& finalPass) const
+    {
+      lno_t numRows = rowmap.extent(0) - 1;
+      lno_t row = clusterVerts(i);
+      size_type rowStart = rowmap(row);
+      size_type rowEnd = rowmap(row + 1);
+      lno_t cluster = vertClusters(row);
+      lno_t clusterStart = clusterOffsets(cluster);
+      //Count the number of entries in this row.
+      //This is how much lcount will be increased by,
+      //yielding the offset corresponding to
+      //these point entries in the cluster entries.
+      lno_t rowEntries = 0;
+      for(size_type j = rowStart; j < rowEnd; j++)
+      {
+        if(edgeMask.test(j))
+          rowEntries++;
+      }
+      if(finalPass)
+      {
+        //if this is the last row in the cluster, update the upper bound in clusterRowmap
+        if(i == clusterStart)
+        {
+          clusterRowmap(cluster) = lcount;
+        }
+        lno_t clusterEdge = lcount;
+        //populate clusterEntries for these edges
+        for(size_type j = rowStart; j < rowEnd; j++)
```

Comment:
Parallel reduce TVR?

Reply:
@lucbv It would be, but it's inside a scan so I can't make it TeamPolicy. It would be nice to have a team+thread level scan with inner vector level.